### PR TITLE
Smartglass

### PR DESCRIPTION
**Pull request by SonixApache**
smartglass
General idea:
https://streamable.com/rk43v
Applying to windows:
https://streamable.com/x7uit
Button functionality:
https://streamable.com/qxj3x
Windoors:
https://streamable.com/u6td8

### Functionality:
- Smartglass can be ~~powered~~ toggled to be transparent, else is opaque
- Add wired glass (glass + wire) to any window or windoor to convert to smartglass
- Use multitool menu to get/set ID tag for buttons - the command is `toggle_transparency`
- Use multitool menu to toggle transparent/opaque
- Unanchor a window to remove the wired glass from it

🆑 
 - rscadd: Added smartglass; glass that can be digitally switched between transparent and opaque. Apply a wired glass tile to a window or windoor to create it, then multitool it for setup.

### CODE REVIEW

**Comment by Duny-** [comment]
Button functionality WHEN

**Comment by SonixApache** [comment]
Gonna work on it more tomorrow or so, whenever my friends don't want to play videogames anymore.
~~Weird how it fails on travis even though it compiled locally, though.~~

File: vgstation13.dme [hunk 1]
```diff
@@ -2222,7 +2223,7 @@
 #include "interface\web\interface.dms"
 #include "maprendering\maprendering.dm"
 #include "maps\_map.dm"
-#include "maps\tgstation.dm"
+#include "maps\test_tiny.dm"
```
Comment (SonixApache) [inline vgstation13.dme]: whoops

**Comment by Ratvar** [comment]
add smartwalls

**Comment by SonixApache** [comment]
you have to go back

**Comment by ghost** [comment]
Cool

File: code/game/machinery/smartglass.dm [hunk 2]
```diff
@@ -0,0 +1,38 @@
+/////////////////////////////////////////
+// Machine that makes glass 'smart'
+// Held within a pane of glass
+// used for linking to buttons 'n shit
+// made when a floor light is attached to a window
+/////////////////////////////////////////
+
+/obj/machinery/smartglass_electronics
+	name = "internal smartglass electronics"
+	desc = "This inside a pane of smart glass. How are you seeing this?"
```
Comment (Intigracy) [inline code/game/machinery/smartglass.dm]: This missing a word.

File: code/game/objects/structures/window.dm [hunk 3]
```diff
@@ -309,10 +318,29 @@ var/list/one_way_windows
 			one_way_windows = list()
 		one_way_windows.Add(src)
 		P.use(1)
-		to_chat(user, "<span class='notice'>You place a sheet of plastic over the window.</span>")
+		to_chat(user, "<span class='notice'>You place a sheet of plastic over the [src].</span>")
```
Comment (Intigracy) [inline code/game/objects/structures/window.dm]: \the [src] or you're gonna get the the window

File: code/game/objects/structures/window.dm [hunk 4]
```diff
@@ -309,10 +318,29 @@ var/list/one_way_windows
...
 		return
 
+	
+	if(istype(W, /obj/item/stack/light_w))
+		var/obj/item/stack/light_w/LT = W
+		if (smartwindow)
+			to_chat(user, "<span class='notice'>This [src] already has electronics in it.</span>")
+			return
+		LT.use(1)
+		to_chat(user, "<span class='notice'>You add some electronics to the [src].</span>")
```
Comment (Intigracy) [inline code/game/objects/structures/window.dm]: the the window

File: code/game/objects/structures/window.dm [hunk 5]
```diff
@@ -309,10 +318,29 @@ var/list/one_way_windows
...
-		to_chat(user, "<span class='notice'>You place a sheet of plastic over the window.</span>")
+		to_chat(user, "<span class='notice'>You place a sheet of plastic over the [src].</span>")
 		overlays += oneway_overlay
 		return
 
+	
+	if(istype(W, /obj/item/stack/light_w))
+		var/obj/item/stack/light_w/LT = W
+		if (smartwindow)
+			to_chat(user, "<span class='notice'>This [src] already has electronics in it.</span>")
```
Comment (Intigracy) [inline code/game/objects/structures/window.dm]: This the window

**Review by Intigracy — DISMISSED** [review]

File: code/game/objects/structures/window.dm (hunk 3 above)
Comment (Exxion) [reply code/game/objects/structures/window.dm]: I still have a hard time understanding how they could have possibly thought that was a good idea
"Hey, let's add text macros to add the to improper nouns, etc.!"
"Better idea: We do that, but also make them present by default, thus making adding words to your shit an opt-out procedure, as well as making several of those macros completely superfluous!"

File: code/game/objects/structures/window.dm (hunk 5 above)
Comment (SonixApache) [reply code/game/objects/structures/window.dm]: There's no `\this`,  right?

**Comment by SonixApache** [comment]
- Fixed grammar
- Added a toggle proc for the one-way windows
- Added a machine inside the glass that handles the multitool menu
- Menu options don't actually work because 1: broken procs and 2: z-level issues
- Added a chat notice when you're out of range when using multitools

**Comment by SonixApache** [comment]
- made radio functionality work
- multitool commands work
- added examine text for smartglass and one-way windows

**Comment by SonixApache** [comment]
- Fixed the one-way toggle button on the multitool menu
- Limited the access range to one tile, like most machinery

Only issue I have now is that it doesn't actually draw power from APCs, and thus losing power doesn't toggle the windows back into opaque mode.

**Comment by SonixApache** [comment]
Added functionality to windoors.
https://puu.sh/xvEqp/8d8107db53.mp4

**Comment by SonixApache** [comment]
Added functionality to windoors.
https://streamable.com/u6td8
Can also remove smartglass functionality by dislocating a window.

**Comment by SonixApache** [comment]
Two things left to fix, added to OP.

File: code/game/objects/structures/window.dm (hunk 5 above)
Comment (Intigracy) [reply code/game/objects/structures/window.dm]: \the, \The // insert 'the' (if needed)
\a, \an,   // insert 'a(n)' or 'some' (if needed)
\A, \An

http://www.byond.com/docs/ref/info.html#/DM/text/macros

**Comment by CptWad** [comment]
ho ho holy shit
I have excellent ideas for this

**Comment by SonixApache** [comment]
Running into a lotta issues with making the internal machinery operate on power. Might just drop those features if I can't figure it out this week.

Alternatively I'd have to make smartglass rely on wire knots, which'd be pretty lame.

**Comment by Exxion** [comment]
I believe all you have to do is manually update the contained machinery's `areaMaster`

**Comment by SonixApache** [comment]
Right now the power functionality is broken regardless. Need to figure that part out first.
Tried manually adding it to the `areaMaster` before, it resulted in the machinery being teleported outta the glass and into a random spot in the room.

**Comment by Exxion** [comment]
No, change the machinery's `areaMaster` var

```
/obj/machinery/smartglass_electronics/use_power()
    areaMaster = GLASS.areaMaster
    return ..()
```
Except replace the spaces with tabs

**Comment by SonixApache** [comment]
What the hell? I thought `use_power` was a var, not a proc?
I mean I getcha on the rest, will definitely implement, but still.

**Comment by Exxion** [comment]
It's both

**Comment by Exxion** [comment]
Alright, we don't really have any var capitalization conventions
But all-caps is like the one format we don't have
All-caps is generally used only for defines

**Comment by Intigracy** [comment]
I wish we had standardization too but what can ya do

File: code/game/machinery/doors/windowdoor.dm [hunk 6]
```diff
@@ -20,12 +20,14 @@
...
+	var/obj/machinery/smartglass_electronics/smartwindow
+	var/was_opaque
 
 /obj/machinery/door/window/New()
 	..()
-	if ((istype(src.req_access) && src.req_access.len) || istext(req_access))
-		src.icon_state = "[src.icon_state]"
-		src.base_state = src.icon_state
+	if ((istype(req_access) && req_access.len) || istext(req_access))
+		icon_state = "[icon_state]"
```
Comment (N3X15) [inline code/game/machinery/doors/windowdoor.dm]: what

File: code/game/machinery/doors/windowdoor.dm [hunk 7]
```diff
@@ -100,50 +113,53 @@
...
+	if (operating == 1) //doors can still open when emag-disabled
 		return 0
 	if (!ticker)
 		return 0
-	if(!src.operating) //in case of emag
-		src.operating = 1
-	flick(text("[]opening", src.base_state), src)
+	if(!operating) //in case of emag
+		operating = 1
+	flick(text("[]opening", base_state), src)
```
Comment (N3X15) [inline code/game/machinery/doors/windowdoor.dm]: Dear god how old is this code if it still has text()?!

File: code/game/machinery/smartglass.dm [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+	density = 0
+	use_power = 1
+	idle_power_usage = 1
+	active_power_usage = 50
+	power_channel = ENVIRON
+	machine_flags = MULTITOOL_MENU
+	
+	//Smartglass vars
+	var/smart_transparency = 0
+	var/obj/structure/window/GLASS //Ref to the window we're in
```
Comment (N3X15) [inline code/game/machinery/smartglass.dm]: Don't use ALLCAPS variable names, that's for constants.

**Review by N3X15 — CHANGES_REQUESTED** [review]

File: code/game/machinery/smartglass.dm (hunk 8 above)
Comment (SonixApache) [reply code/game/machinery/smartglass.dm]: roger

**Comment by SonixApache** [comment]
Varname changed to `Ourwindow`.
Reverted all the `src` removal to make the diffs more readable.
Compiled and checked, seems okay.

**Review by N3X15 — APPROVED** [review]

**Pull request merged**